### PR TITLE
[Merged by Bors] - feat(AlgebraicGeometry/EllipticCurve/Projective): implement group operation polynomials for projective coordinates

### PR DESCRIPTION
**Pull request by Multramate**
Define auxiliary polynomials for the secant-and-tangent negation `neg` and addition `add` on `Fin 3 -> F`. Note that the group operations are defined from scratch instead of pulling back the group operations from the affine case, as this will be more convenient for rewriting and allows them to be defined over a ring. I don't see an easy way to define an analogous `slope` function as in `Affine` when there are now `Z`-coordinates in the polynomials.

This is the second in a series of four PRs leading to #8485 

---

- [x] depends on: #9416

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by grunweg** [comment]
This looks like a great PR, which is not on the review queue any more because it incurred a merge conflict.
Can you merge master, please? Thanks!

**Comment by Multramate** [comment]
> This looks like a great PR, which is not on the review queue any more because it incurred a merge conflict. Can you merge master, please? Thanks!

Thanks, but note that this should only be merged after #14210 is merged, and that PR is basically ready to be merged.

File: Mathlib/AlgebraicGeometry/EllipticCurve/Projective.lean
Comment (kbuzzard) [inline Mathlib/AlgebraicGeometry/EllipticCurve/Projective.lean:242]: There are now two `hP`s. Isn't that a bit confusing? Or is it intentional?

File: Mathlib/AlgebraicGeometry/EllipticCurve/Projective.lean
Comment (kbuzzard) [inline Mathlib/AlgebraicGeometry/EllipticCurve/Projective.lean:889]: This lemma really confused me, because it's not true that the Z coordinate of P+P is zero. Maybe add a docstring explaining that `addZ` is only valid for distinct points and so this lemma is just explaining a junk value, and that one should use `dblZ` instead when the points are the same? (assuming that's what's going on!)

File: Mathlib/AlgebraicGeometry/EllipticCurve/Projective.lean
Comment (kbuzzard) [inline Mathlib/AlgebraicGeometry/EllipticCurve/Projective.lean:903]: Maybe somewhere (docstring? module docstring?) it should be mentioned that `X_eq` means "X coordinates of P and Q are equal once normalised by dividing by Z coordinates"? Or maybe not -- I guess I just figured this out by looking at the statement.

File: Mathlib/AlgebraicGeometry/EllipticCurve/Projective.lean
Comment (kbuzzard) [inline Mathlib/AlgebraicGeometry/EllipticCurve/Projective.lean:991]: I think "negated addition" sounds a bit weird/complicated. Maybe just say "negated addition `-(P+Q)` of two distinct..." if that's what it is? This time it's impossible to tell from the code :-) Maybe also say that this produces a junk value of 0 if P=Q?

File: Mathlib/AlgebraicGeometry/EllipticCurve/Projective.lean
Comment (kbuzzard) [inline Mathlib/AlgebraicGeometry/EllipticCurve/Projective.lean:1109]: Do you understand why you need `erw` here whereas for most of the PR `rw` has been fine?

File: Mathlib/AlgebraicGeometry/EllipticCurve/Projective.lean
Comment (kbuzzard) [inline Mathlib/AlgebraicGeometry/EllipticCurve/Projective.lean:1070]: In all of these definitions I would add "returns a junk value if P=Q" just to be clear that this is what's happening (I was confused at first when reviewing this PR, so maybe others will be confused when reading it). My confusion stems from the fact that I was incorrectly assuming that in projective coordinates you can get away with one formula for addition, but apparently this is not correct. I guess that in affine coordinates there are many formulae (P+0, P+P, 0+P, P+(-P) etc) so presumably there are still gains to be had with two formulae.

File: Mathlib/AlgebraicGeometry/EllipticCurve/Projective.lean
Comment (kbuzzard) [inline Mathlib/AlgebraicGeometry/EllipticCurve/Projective.lean:1114]: Why not just `add` but in the appropriate namespace so people know it's an elliptic curve add?

**Review by kbuzzard — APPROVED** [review]
This is a stellar effort, well done. I think the large formulae are inevitable; formulae for addition and doubling are typically presented in the textbooks for curves of the form y^2=x^3+Ax+B but because you're (correctly) doing the general case I think it's inevitable that large formulae will appear.

File: Mathlib/AlgebraicGeometry/EllipticCurve/Projective.lean
Comment (Multramate) [reply Mathlib/AlgebraicGeometry/EllipticCurve/Projective.lean:242]: Probably fine? The first `hP` doesn't appear inside the `have`, and the second `hP` doesn't appear outside the `have`.

File: Mathlib/AlgebraicGeometry/EllipticCurve/Projective.lean
Comment (Multramate) [reply Mathlib/AlgebraicGeometry/EllipticCurve/Projective.lean:903]: I will remember to add these into the module docstrings when I refactor them into a separate file.

File: Mathlib/AlgebraicGeometry/EllipticCurve/Projective.lean
Comment (Multramate) [reply Mathlib/AlgebraicGeometry/EllipticCurve/Projective.lean:1109]: Yeah this is interesting. I really try to avoid `erw` whenever possible, but there are a couple of instances where I need to rewrite something like `a * ![0, 1, 0] 2`, and it's just faster to do `erw [mul_zero]` to rewrite `a * 0 = ...`. This lemma is similar, but instead I need to rewrite `W.negY ![a, b, c] / c` using a lemma that says `W.negY ![a, b, c] / ![a, b, c] 2 = ...`.

File: Mathlib/AlgebraicGeometry/EllipticCurve/Projective.lean
Comment (Multramate) [reply Mathlib/AlgebraicGeometry/EllipticCurve/Projective.lean:1114]: This is not the addition of points, rather the resulting tuple of the addition of point representatives, which is mathematically correct if and only if the two points are distinct, so it doesn't deserve its own namespace. There will be a `def add (P Q : Fin 3 → R) : Fin 3 → R := if P ≈ Q then W'.dblXYZ P else W'.addXYZ P Q` in an upcoming PR that addresses the issue of two points being distinct, and there will be a `def Point.add (P Q : W.Point) : W.Point` in a subsequent PR.

File: Mathlib/AlgebraicGeometry/EllipticCurve/Projective.lean
Comment (Multramate) [reply Mathlib/AlgebraicGeometry/EllipticCurve/Projective.lean:889]: Thanks, I will add this to the definition of `addZ`.

File: Mathlib/AlgebraicGeometry/EllipticCurve/Projective.lean
Comment (Multramate) [reply Mathlib/AlgebraicGeometry/EllipticCurve/Projective.lean:1070]: Yeah, thanks! I will add this too.

File: Mathlib/AlgebraicGeometry/EllipticCurve/Projective.lean
Comment (Multramate) [reply Mathlib/AlgebraicGeometry/EllipticCurve/Projective.lean:991]: Sure, I will add this.

**Comment by kbuzzard** [comment]
maintainer merge

File: Mathlib/AlgebraicGeometry/EllipticCurve/Projective.lean
Comment (riccardobrasca) [inline Mathlib/AlgebraicGeometry/EllipticCurve/Projective.lean:469]: You can write `variable {P Q : Fin 3 → F}`

File: Mathlib/AlgebraicGeometry/EllipticCurve/Projective.lean
Comment (riccardobrasca) [inline Mathlib/AlgebraicGeometry/EllipticCurve/Projective.lean:553]: Same as above.

**Review by riccardobrasca — APPROVED** [review]
LGTM, thanks!

Can you add a bit of docstring at the beginning of the file explaining where to find the formulas?

bors d+

File: Mathlib/AlgebraicGeometry/EllipticCurve/Projective.lean
Comment (Multramate) [reply Mathlib/AlgebraicGeometry/EllipticCurve/Projective.lean:469]: Some of these only require the ring to be commutative (possibly with `NoZeroDivisors`), while others need the ring to be a field, so I thought the variables `P` and `Q` should be instantiated along with the lemma.

File: Mathlib/AlgebraicGeometry/EllipticCurve/Projective.lean
Comment (alreadydone) [inline Mathlib/AlgebraicGeometry/EllipticCurve/Projective.lean:683]: Note to myself: 
`z^2  (negDblY - y polynomialY^3 - c1 polynomialY) + polynomialX^3 = c2 polynomial`,
where
```
c1 = -3  a1^2  x^2  y - 12 a1 x y^2 - 18 y^3 - a1 a2^2 x^2 z + 
 3 a1 a4 x^2 z - 4 a2^2 x y z - 6 a1 a3 x y z + 12 a4 x y z + 
 a1^3 y^2 z + 5 a1 a2 y^2 z - 27 a3 y^2 z - 2 a2^2 a3 x z^2 + 
 6 a3 a4 x z^2 + 4 a1 a2 a3 y z^2 - 9 a3^2 y z^2 - 2 a1^2 a4 y z^2 - 
 2 a2 a4 y z^2 + 18 a6 y z^2 - a2 a3 a4 z^3 + a1 a4^2 z^3 - 
 3 a1 a2 a6 z^3 + 9 a3  a6  z^3
c2 = 3 (9 x^3 + 9 a2 x^2 z + 9 y^2 z - a1^2 a2 x z^2 + 3 a1 a3 x z^2 - 
   2 a1 a2 y z^2 + 12 a3 y z^2 - a1 a2 a3 z^3 + 3 a3^2 z^3)
```

File: Mathlib/AlgebraicGeometry/EllipticCurve/Projective.lean
Comment (Multramate) [reply Mathlib/AlgebraicGeometry/EllipticCurve/Projective.lean:613]: Yeah that was the original expression that I called `preDblX` (and `preAddX` for addition). I thought that since we're already having massive polynomials, might as well write them out in full. Is there a use for `preDblX` by itself?

Comment (alreadydone) [reply Mathlib/AlgebraicGeometry/EllipticCurve/Projective.lean:613]: Well your `dblZ` is still in factored form. Why not factor `dblX` as well to make it shorter?

File: Mathlib/AlgebraicGeometry/EllipticCurve/Projective.lean
Comment (alreadydone) [reply Mathlib/AlgebraicGeometry/EllipticCurve/Projective.lean:683]: Better: the cubic factor in `dblX` can be replaced by 
```
-a2 x^3 + a1 x^2 y + x y^2 - a1 a3 x^2 z - 3 a4 x^2 z + a3 x y z - 
 2 a3^2 x z^2 - 9 a6 x z^2 - a2 a3^2 z^3 + a1 a3 a4 z^3 + a4^2 z^3 - 
 a1^2 a6 z^3 - 4 a2 a6 z^3
```
which is
`x polynomialY^2 - ψh₃ - 3 (x/z) polynomial`.
Now trying to do the same for `negDblY` ...

File: Mathlib/AlgebraicGeometry/EllipticCurve/Projective.lean
Comment (alreadydone) [reply Mathlib/AlgebraicGeometry/EllipticCurve/Projective.lean:613]: Better: the cubic factor in `dblX` can be replaced by 
```
-a2 x^3 + a1 x^2 y + x y^2 - a1 a3 x^2 z - 3 a4 x^2 z + a3 x y z - 
 2 a3^2 x z^2 - 9 a6 x z^2 - a2 a3^2 z^3 + a1 a3 a4 z^3 + a4^2 z^3 - 
 a1^2 a6 z^3 - 4 a2 a6 z^3
```
which is
`x polynomialY^2 - ψh₃ - 3 (x/z) polynomial`.
Now trying to do the same for `negDblY` ...

File: Mathlib/AlgebraicGeometry/EllipticCurve/Projective.lean
Comment (alreadydone) [reply Mathlib/AlgebraicGeometry/EllipticCurve/Projective.lean:683]: Originally I worked out this formula in order to show when you apply `negDblY` to a 2-torsion nonsingular point (`polynomialY = 0` but `polynomialX ≠ 0`) then `negDblY ≠ 0`, so our doubling formula still holds in the P+P=O case. Now, with my accidental discovery that `negDblY` can be taken to be `Y polynomialY^3 - ψh₃ polynomialX` we can give an alternative proof: at a 2-torsion nonsingular point P ≠ O we have `ψh₃ ≠ 0` because otherwise P is also 3-torsion and therefore P = O; since polynomialY=0 and polynomialX≠0 we conclude negDblY≠0. This is a bit circular though, as it uses a high-powered result #13782, even though the PR uses only Jacobian API which is independent of the projective case we consider here.

**Pull request closed without merging**
